## Importação de bibliotecas necessárias

In [1]:
import pandas as pd
import numpy as np

## Carregando o dado

In [3]:
df = pd.read_json('Telco-Customer-Churn.json')
df.head(5)

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

## Explorando o dado

In [4]:
df.customer[0]

{'gender': 'Female',
 'SeniorCitizen': 0,
 'Partner': 'Yes',
 'Dependents': 'Yes',
 'tenure': 9}

In [5]:
df.phone[0]

{'PhoneService': 'Yes', 'MultipleLines': 'No'}

In [6]:
df.internet[0]

{'InternetService': 'DSL',
 'OnlineSecurity': 'No',
 'OnlineBackup': 'Yes',
 'DeviceProtection': 'No',
 'TechSupport': 'Yes',
 'StreamingTV': 'Yes',
 'StreamingMovies': 'No'}

In [7]:
df.account[0]

{'Contract': 'One year',
 'PaperlessBilling': 'Yes',
 'PaymentMethod': 'Mailed check',
 'Charges': {'Monthly': 65.6, 'Total': '593.3'}}

## Normalização dos dados

In [8]:
df_customer = pd.json_normalize(df.customer)
df_customer.head(5)

gender  SeniorCitizen Partner Dependents  tenure
0  Female              0     Yes        Yes       9
1    Male              0      No         No       9
2    Male              0      No         No       4
3    Male              1     Yes         No      13
4  Female              1     Yes         No       3

In [9]:
df_phone = pd.json_normalize(df.phone)
df_phone.head(5)

PhoneService MultipleLines
0          Yes            No
1          Yes           Yes
2          Yes            No
3          Yes            No
4          Yes            No

In [10]:
df_internet = pd.json_normalize(df.internet)
df_internet.head(5)

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   
3     Fiber optic             No          Yes              Yes          No   
4     Fiber optic             No           No               No         Yes   

  StreamingTV StreamingMovies  
0         Yes              No  
1          No             Yes  
2          No              No  
3         Yes             Yes  
4         Yes              No

In [11]:
df_account = pd.json_normalize(df.account)
df_account.head(5)

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0        One year              Yes      Mailed check             65.6   
1  Month-to-month               No      Mailed check             59.9   
2  Month-to-month              Yes  Electronic check             73.9   
3  Month-to-month              Yes  Electronic check             98.0   
4  Month-to-month              Yes      Mailed check             83.9   

  Charges.Total  
0         593.3  
1         542.4  
2        280.85  
3       1237.85  
4         267.4

## Unindo as tabelas

In [22]:
dados = df[list(df.columns[:2])]
dados.head(5)

customerID Churn
0  0002-ORFBO    No
1  0003-MKNFE    No
2  0004-TLHLJ   Yes
3  0011-IGKFF   Yes
4  0013-EXCHZ   Yes

In [25]:
tabelas = [df_customer, df_phone, df_internet, df_account]

for tabela in tabelas:
    concat_df = tabela
    dados = pd.concat([dados, concat_df], axis = 1)

dados.head(5)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

## Sumarização do dado

In [26]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [32]:
for column in dados.columns:
    print(f'---- {column} ----')
    print(dados[column].value_counts())
    print('')

---- customerID ----
0002-ORFBO    1
6614-VBEGU    1
6637-KYRCV    1
6635-MYYYZ    1
6635-CPNUN    1
             ..
3374-TTZTK    1
3374-PZLXD    1
3374-LXDEV    1
3373-YZZYM    1
9995-HOTOH    1
Name: customerID, Length: 7267, dtype: int64

---- Churn ----
No     5174
Yes    1869
        224
Name: Churn, dtype: int64

---- gender ----
Male      3675
Female    3592
Name: gender, dtype: int64

---- SeniorCitizen ----
0    6085
1    1182
Name: SeniorCitizen, dtype: int64

---- Partner ----
No     3749
Yes    3518
Name: Partner, dtype: int64

---- Dependents ----
No     5086
Yes    2181
Name: Dependents, dtype: int64

---- tenure ----
1     634
72    369
2     246
3     207
4     185
     ... 
38     60
39     59
44     54
36     50
0      11
Name: tenure, Length: 73, dtype: int64

---- PhoneService ----
Yes    6560
No      707
Name: PhoneService, dtype: int64

---- MultipleLines ----
No                  3495
Yes                 3065
No phone service     707
Name: MultipleLines, dtype: i

#### A coluna 'Churn' apresenta 224 valores vazios que devem ser apagados

In [36]:
dados = dados[dados['Churn'] != '']
dados.reset_index(drop = True, inplace = True)

In [40]:
for column in dados.columns:
    print(f'---- {column} ----')
    print(dados[column].value_counts())
    print('')

---- customerID ----
0002-ORFBO    1
6616-AALSR    1
6625-UTXEW    1
6625-IUTTT    1
6625-FLENO    1
             ..
3352-RICWQ    1
3352-ALMCK    1
3351-NQLDI    1
3351-NGXYI    1
9995-HOTOH    1
Name: customerID, Length: 7043, dtype: int64

---- Churn ----
No     5174
Yes    1869
Name: Churn, dtype: int64

---- gender ----
Male      3555
Female    3488
Name: gender, dtype: int64

---- SeniorCitizen ----
0    5901
1    1142
Name: SeniorCitizen, dtype: int64

---- Partner ----
No     3641
Yes    3402
Name: Partner, dtype: int64

---- Dependents ----
No     4933
Yes    2110
Name: Dependents, dtype: int64

---- tenure ----
1     613
72    362
2     238
3     200
4     176
     ... 
28     57
39     56
44     51
36     50
0      11
Name: tenure, Length: 73, dtype: int64

---- PhoneService ----
Yes    6361
No      682
Name: PhoneService, dtype: int64

---- MultipleLines ----
No                  3390
Yes                 2971
No phone service     682
Name: MultipleLines, dtype: int64

---- I

In [45]:
idx = dados[dados['Charges.Total'] == ' '].index
dados.loc[idx, 'Charges.Total'] = dados.loc[idx, 'Charges.Monthly']